File: python/307_Range Sum Query - Mutable.py
#!/usr/bin/env python2
'''
class NumArray(object):

    def __init__(self, nums):
        """
        :type nums: List[int]
        """
        if not nums or len(nums) == 0:
            self.nums = []
            self.store = []
        else:
            self.nums = nums
            self.store = [nums[0]]
            for num in nums[1:]:
                self.store.append(num + self.store[-1])

    def update(self, i, val):
        """
        :type i: int
        :type val: int
        :rtype: void
        """
        diff = val - self.nums[i]
        self.nums[i] = val
        for j in range(i, len(self.store)):
            self.store[j] += diff


    def sumRange(self, i, j):
        """
        :type i: int
        :type j: int
        :rtype: int
        """
        return self.store[j] - (self.store[i-1] if i > 0 else 0)
'''
class NumArray(object):

    def __init__(self, nums):
        """
        :type nums: List[int]
        """
        if not nums or len(nums) == 0:
            return
        self.root = self.build_tree(0, len(nums)-1, nums)

    def update(self, i, val):
        """
        :type i: int
        :type val: int
        :rtype: void
        """
        self.updateTree(self.root, i, val)

    def sumRange(self, i, j):
        """
        :type i: int
        :type j: int
        :rtype: int
        """
        return self.getSum(self.root,i, j)

    class TreeNode(object):
        def __init__(self, left, right, sum, left_child, right_child):
            self.left = left
            self.right = right
            self.sum = sum
            self.left_child = left_child
            self.right_child = right_child

    def build_tree(self, start, end, nums):
        if start == end:
            node = self.TreeNode(start, end, nums[start], None, None)
        else:
            mid = (start + end) / 2
            left_child = self.build_tree(start, mid, nums)
            right_child = self.build_tree(mid+1, end, nums)
            sum = left_child.sum + right_child.sum
            node = self.TreeNode(start, end, sum, left_child, right_child)
        return node

    def getSum(self, root, i, j):
        if not root:
            return 0
        if root.left == i and root.right == j:
            return root.sum
        mid = (root.left + root.right) / 2
        if mid >= j:
            return self.getSum(root.left_child, i, j)
        elif mid < i:
            return self.getSum(root.right_child, i, j)
        else:
            return self.getSum(root.left_child, i, mid) + self.getSum(root.right_child, mid+1, j)

    def updateTree(self, root, i, val):
        if not root:
            return
        if root.left == root.right:
            root.sum = val
            return
        mid = (root.left + root.right) / 2
        if i <= mid:
            self.updateTree(root.left_child, i, val)
        else:
            self.updateTree(root.right_child, i, val)
        root.sum = root.left_child.sum + root.right_child.sum


# Your NumArray object will be instantiated and called as such:
obj = NumArray([0, 9, 5, 7 ,3])
print obj.sumRange(4, 4)
print obj.sumRange(2, 4)

#print obj.sumRange()
#obj.update()

File: java/src/brand_new/_307_Range_Sum_Query_Mutable.java
package brand_new;

public class _307_Range_Sum_Query_Mutable {

	public static void main(String[] args) {
		NumArray na = new NumArray(new int[]{1,3,5});
		System.out.println(na.sumRange(0,2));
		na.update(1, 2);
		System.out.println(na.sumRange(0, 2));
		
	}

}

class NumArray {
	
	class SegmentTreeNode{
		int val;
		int start;
		int end;
		SegmentTreeNode left, right;
		
		public SegmentTreeNode(int start, int end){
			this.start = start;
			this.end = end;
		}
	} 

	SegmentTreeNode root = null;
	
    public NumArray(int[] nums) {
    	root = building(nums, 0, nums.length-1);
    }
    
    public void update(int i, int val) {
    	update(root, i, val);
    }
    
    public int sumRange(int i, int j) {
    	return sumRange(root, i, j);
    }
    
    public SegmentTreeNode building(int[] nums, int start, int end){
    	if(start > end){
    		return null;
    	}
    	
    	SegmentTreeNode node = new SegmentTreeNode(start, end);
    	
    	if(start == end){
    		node.val = nums[start];
    	}else{
    		int mid = start + (end - start) / 2;
    		node.left = building(nums, start, mid);
    		node.right = building(nums, mid+1, end);
    		node.val = node.left.val + node.right.val;
    	}
    	
    	return node;
    }
    
    public void update(SegmentTreeNode node, int i, int val){
    	if(node.start == node.end){
    		node.val = val;
    		return;
    	}
   
    	int mid = node.start + (node.end - node.start) / 2;
    	if(i <= mid){
    		update(node.left, i, val);
    	}else{
    		update(node.right, i, val);
    	}
    	
    	node.val = node.left.val + node.right.val;
    }
    
    public int sumRange(SegmentTreeNode node, int i, int j){
    	if(node.start == i && node.end == j)
    		return node.val;
    	
    	int mid = node.start + (node.end - node.start) / 2;
    	if(j <= mid){
    		return sumRange(node.left, i, j);
    	}else if(i > mid){
    		return sumRange(node.right, i, j);
    	}else{
    		return sumRange(node.left, i, mid) + sumRange(node.right, mid+1, j);
    	}
    }
}